### PR TITLE
[one-cmds] Integrate 'circle-mpqsolver' with 'onecc'

### PR DESCRIPTION
**Pull request by stamalakhov**
This commit implements integration of 'circle-mpqsolver' with 'onecc'.

for sample [one-quantize] section
```
[one-quantize]
input_path=/home/stanislav/repos/WorkSpaces/ampq/mobilenet_v1_1.0_224.opt.circle
output_path=/home/stanislav/repos/WorkSpaces/ampq/model.q_opt.circle
input_data=/mnt/storage/datasets/ms_coco/dataset_MB_v1_10.h5
ampq=True
verbose=True
save_intermediate=True
ampq_qerror_ratio=0.1
ampq_algorithm=bisection
bisection_type=auto
```
it produced the following log for `circle-mpqsolver`:

```
Recording 0'th data
Recording 1'th data
Recording 2'th data
Recording 3'th data
Recording 4'th data
Recording 5'th data
Recording 6'th data
Recording 7'th data
Recording 8'th data
Recording 9'th data
Recording finished. Number of recorded data: 10
>> Searching mixed precision configuration 
model:/home/stanislav/repos/WorkSpaces/ampq/mobilenet_v1_1.0_224.opt.minmax_recorded.circle
dataset: /mnt/storage/datasets/ms_coco/dataset_MB_v1_10.h5
input dtype: uint8
output dtype: uint8
algorithm: bisection (auto)
qerror metric: MAE
target qerror ratio: 0.1

>> Computing baseline qerrors
Full int16 model qerror: 9.27323e-05
Full uint8 model qerror: 0.000254375
Target qerror: 0.000108897

>> Running bisection(auto) algorithm
Qerror of front half: 147884
Qerror of rear half: 45008.1
Front part will be Q16, while the rear will be Q8

Looking for the optimal configuration in [0 , 32] depth segment
Qerror at depth 16 is 0.000130434 > target qerror (0.000108897)
Looking for the optimal configuration in [16 , 32] depth segment
Qerror at depth 24 is 0.000107807 < target qerror (0.000108897)
Looking for the optimal configuration in [16 , 24] depth segment
Qerror at depth 20 is 0.000124079 > target qerror (0.000108897)
Looking for the optimal configuration in [20 , 24] depth segment
Qerror at depth 22 is 0.000130697 > target qerror (0.000108897)
Looking for the optimal configuration in [22 , 24] depth segment
Qerror at depth 23 is 0.000113433 > target qerror (0.000108897)
Found the best configuration at depth 24
Saving output model to /home/stanislav/repos/WorkSpaces/ampq/model.q_opt.circle

```
[model.q_opt.circle.log](https://github.com/Samsung/ONE/files/12268730/model.q_opt.circle.log)

Related: #11146
ONE-DCO-1.0-Signed-off-by: s.malakhov <s.malakhov@partner.samsung.com>

### CODE REVIEW

**Comment by stamalakhov** [comment]
@jinevening 
`save_intermediate`
and `ampq_bisection_visq` options are  also supported.

File: compiler/one-cmds/one-quantize [hunk 1]
```diff
@@ -226,6 +226,28 @@ def _get_parser():
...
+
+    # ampq_qerror_ratio
+    ampq_quant_group.add_argument(
+        '--ampq_qerror_ratio', type=str, help='quantization error ratio ([0, 1])')
+
+    # ampq_algorithm
+    ampq_quant_group.add_argument(
+        '--ampq_algorithm',
+        type=str,
+        help='one of bisection_auto, bisection_true, bisection_false')
```
Comment (seanshpark) [inline compiler/one-cmds/one-quantize:243]: Q) why not use something like algorithm name? just `bisection`, ... and whatever name? using `auto`, `true` and `false` seems strange.

**Comment by seanshpark** [comment]
plz update `infra/debian/compiler/one-compiler.install` too

File: compiler/one-cmds/one-quantize (hunk 1 above)
Comment (jinevening) [reply compiler/one-cmds/one-quantize:243]: We can split the arguments, but I'm not sure which one is better.

as-is
```
ampq_algorithm= # bisection_auto, bisection_true, bisection_false
```

split
```
ampq_algorithm=bisection
bisection_type= # auto, true, false
```


@stamalakhov auto seems ok, but `bisection_true` and `bisection_false` look not natural. How about naming `bisection_i16_front` or `bisection_i16_back` ?

Comment (seanshpark) [reply compiler/one-cmds/one-quantize:243]: For me
```
ampq_algorithm=bisection
bisection_type= # auto, true, false
```
looks better.

`isection_i16_front or bisection_i16_back` also looks nice then just `true` or `false`. 
`bisection_false` gives me that `do not use bisection...` and this doesn't look logical.

Comment (jinevening) [reply compiler/one-cmds/one-quantize:243]: Assembling the two options, the below interface looks good to me :)

```
ampq_algorithm=bisection
bisection_type= # auto, i16_front, i16_back
```

**Comment by jinevening** [comment]
Could you add tests under `one-cmds/tests`? We may need at least three `one-quantize` test (for each algorithm), and one `onecc` test (maybe just for auto).

I think you can use `inception_v3.circle` and `inception_v3_test_data.h5` as test materials (please see `prepare_test_materials.sh` for more details)

**Comment by stamalakhov** [comment]
> Could you add tests under `one-cmds/tests`? We may need at least three `one-quantize` test (for each algorithm), and one `onecc` test (maybe just for auto).
> 
> I think you can use `inception_v3.circle` and `inception_v3_test_data.h5` as test materials (please see `prepare_test_materials.sh` for more details)

Ok. Thank you.

**Comment by stamalakhov** [comment]
> inception_v3_test_data.h5

@jinevening 
I've implemented three tests. However one of them (with 'auto' mode) fails with the following output
```
Error raised while running the below command
visq: visq: CalledProcessError: Command '['/home/stanislav/repos/ONE_f_1/ONE/build/release_build/bin/dalgona', '--input_model', './inception_v3.circle', '--analysis', '/home/stanislav/repos/ONE_f_1/ONE/build/release_build/bin/visqlib/DumpFP32FM.py', '--input_data', '/tmp/tmpdj55tmri/inputs.h5', '--analysis_args', '/tmp/tmp76xs00sr']' returned non-zero exit status 255.
/home/stanislav/repos/ONE_f_1/ONE/build/release_build/bin/dalgona --input_model ./inception_v3.circle --analysis /home/stanislav/repos/ONE_f_1/ONE/build/release_build/bin/visqlib/DumpFP32FM.py --input_data /tmp/tmpdj55tmri/inputs.h5 --analysis_args /tmp/tmp76xs00sr
ERROR: Buffer size does not match with the size of tensor data
```
I'll try to investigate that.

**Comment by jinevening** [comment]
> ERROR: Buffer size does not match with the size of tensor data

Maybe you've given a wrong input data. The message means that the input data size does not match with the model input.

I've tested with the below cmd, and it worked.

```
/usr/share/one/bin/visq -f inception_v3.circle -q inception_v3.quantized.circle -d inception_v3_test_data.h5 --dump_dot_graph --tae_output mpeir.visq.json
```

**Comment by stamalakhov** [comment]
> > ERROR: Buffer size does not match with the size of tensor data
> 
> Maybe you've given a wrong input data. The message means that the input data size does not match with the model input.
> 
> I've tested with the below cmd, and it worked.
> 
> ```
> /usr/share/one/bin/visq -f inception_v3.circle -q inception_v3.quantized.circle -d inception_v3_test_data.h5 --dump_dot_graph --tae_output mpeir.visq.json
> ```

Ahh. Ok. Thank you.

**Comment by stamalakhov** [comment]
@jinevening 
Now all tests are fixed.
It occured that 'rawData' mode was not supported for batch mode in `visq` and it treated `uint8` raw input as `float` data which was the cause of the issue. Please see changes in `visq`.

**Comment by seanshpark** [comment]
Plz revisit `infra/scripts/compiler_modules.sh` too

**Comment by stamalakhov** [comment]
> Plz revisit `infra/scripts/compiler_modules.sh` too

ok

**Comment by stamalakhov** [comment]
> Plz revisit `infra/scripts/compiler_modules.sh` too

@seanshpark
done

**Comment by stamalakhov** [comment]
@jinevening @seanshpark 
can i split this draft into PR's or there are some issues?

**Comment by seanshpark** [comment]
> can i split this draft into PR's or there are some issues?

I've none. Plz go as you wish.

**Comment by stamalakhov** [comment]
> I've none. Plz go as you wish.

ok.

**Comment by jinevening** [comment]
> can i split this draft into PR's or there are some issues?

I think it's ok to push PRs.

**Comment by stamalakhov** [comment]
> I think it's ok to push PRs.

ok.

**Comment by stamalakhov** [comment]
Seems like all PR's were merged, so i's safe to close this one.

**Pull request closed without merging**